src/server/pkg/obj/cache_client.go

```diff
@@ -0,0 +1,192 @@
+package obj
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	log "github.com/sirupsen/logrus"
+)
+
+var _ Client = &cacheClient{}
+
+type cacheClient struct {
+	slow, fast Client
+
+	mu           sync.Mutex
+	cache        *simplelru.LRU
+	pathLocks    sync.Map
+	populateOnce sync.Once
+}
+
+// NewCacheClient returns slow wrapped in an LRU write-through cache using fast for storing
+// cached data.
+func NewCacheClient(slow, fast Client, size int) Client {
+	if size == 0 {
+		return slow
+	}
+	client := &cacheClient{
+		slow: slow,
+		fast: fast,
+	}
+	cache, err := simplelru.NewLRU(size, client.onEvicted)
+	if err != nil {
+		// lru.NewWithEvict only errors for size < 1
+		panic(err)
+	}
+	client.cache = cache
+	return client
+}
+
+func (c *cacheClient) Reader(ctx context.Context, p string, offset, size uint64) (io.ReadCloser, error) {
+	c.doPopulateOnce(ctx)
+	if offset > 0 || size > 0 {
```

Comment:
Why does this need to go through the slow client? We cannot get part of a chunk from the fast client? If we end up leaning fully on the cache client for lexicographical reads (rather than caching a chunk in the chunk reader), then we are going to want these types of reads to go to the fast client.

Reply:
Right now the storage layer only gets full chunks so that path is not used.  We don't want to cache pieces of chunks, that greatly complicates things, so in the case of a cache miss we want to get the whole thing from slow, and then get the fragment from fast. I don't think that would be too difficult to add. I can look into it if we pick this vs. the object caching.

Reply:
Yeah, it would not be used now, but I could see it being used later. Also, I'm not suggesting caching pieces of chunks. I am suggesting being able to read parts of full chunks that are cached. Our access patterns are lexicographical, so caching parts of chunks would probably be less performant even if we implemented it well.

Reply:
Actually if we always gzip the chunks, it might never be used because you can't seek into a gzip stream.

Reply:
This should work now.